File: android/test-app-java/src/main/java/com/arthenica/ffmpegkit/test/MainActivity.java
/*
 * Copyright (c) 2018-2026 Taner Sener
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package com.arthenica.ffmpegkit.test;

import android.graphics.Color;
import android.graphics.Typeface;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.util.Log;
import android.view.View;
import android.widget.HorizontalScrollView;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.activity.EdgeToEdge;
import androidx.activity.SystemBarStyle;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.content.ContextCompat;
import androidx.core.graphics.Insets;
import androidx.core.view.ViewCompat;
import androidx.core.view.WindowInsetsCompat;
import androidx.viewpager2.widget.ViewPager2;

import com.arthenica.ffmpegkit.FFmpegKit;
import com.arthenica.ffmpegkit.FFmpegKitConfig;
import com.arthenica.ffmpegkit.FFmpegSession;
import com.arthenica.ffmpegkit.FFprobeKit;
import com.arthenica.ffmpegkit.FFprobeSession;
import com.arthenica.ffmpegkit.Level;
import com.arthenica.ffmpegkit.Signal;
import com.arthenica.ffmpegkit.util.ResourcesUtil;
import com.arthenica.smartexception.java.Exceptions;

import java.io.File;
import java.io.IOException;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;

public class MainActivity extends AppCompatActivity {

    public static final String TAG = "ffmpeg-kit-next-test";

    static {
        Exceptions.registerRootPackage("com.arthenica");
    }

    protected static final Handler handler = new Handler(Looper.getMainLooper());

    @Override
    protected void onCreate(final Bundle savedInstanceState) {
        EdgeToEdge.enable(this, SystemBarStyle.dark(Color.TRANSPARENT));

        super.onCreate(savedInstanceState);

        setContentView(R.layout.activity_main);

        final ViewPager2 viewPager = findViewById(R.id.pager);
        final PagerAdapter pagerAdapter = new PagerAdapter(this, this);
        viewPager.setAdapter(pagerAdapter);
        setupTabs(viewPager, pagerAdapter);

        final View toolbar = findViewById(R.id.toolbar);
        final View rootLayout = findViewById(R.id.rootLayout);
        ViewCompat.setOnApplyWindowInsetsListener(rootLayout, (v, windowInsets) -> {
            final Insets bars = windowInsets.getInsets(
                    WindowInsetsCompat.Type.systemBars() | WindowInsetsCompat.Type.displayCutout());
            toolbar.setPadding(0, bars.top, 0, 0);
            viewPager.setPadding(bars.left, 0, bars.right, bars.bottom);
            return WindowInsetsCompat.CONSUMED;
        });

        try {
            registerApplicationFonts();
            Log.d(TAG, "Application fonts registered.");
        } catch (final IOException e) {
            Log.e(TAG, String.format("Font registration failed.%s.", Exceptions.getStackTraceString(e)));
        }

        Log.d(TAG, "Listing supported camera ids.");
        listSupportedCameraIds();

        FFmpegKitConfig.ignoreSignal(Signal.SIGXCPU);
        FFmpegKitConfig.setLogLevel(Level.AV_LOG_INFO);
    }

    public static void listFFmpegSessions() {
        List<FFmpegSession> ffmpegSessions = FFmpegKit.listSessions();
        Log.d(TAG, "Listing FFmpeg sessions.");
        for (int i = 0; i < ffmpegSessions.size(); i++) {
            FFmpegSession session = ffmpegSessions.get(i);
            Log.d(TAG, String.format("Session %d = id:%d, startTime:%s, duration:%s, state:%s, returnCode:%s.",
                    i,
                    session.getSessionId(),
                    session.getStartTime(),
                    session.getDuration(),
                    session.getState(),
                    session.getReturnCode()));
        }
        Log.d(TAG, "Listed FFmpeg sessions.");
    }

    public static void listFFprobeSessions() {
        List<FFprobeSession> ffprobeSessions = FFprobeKit.listFFprobeSessions();
        Log.d(TAG, "Listing FFprobe sessions.");
        for (int i = 0; i < ffprobeSessions.size(); i++) {
            FFprobeSession session = ffprobeSessions.get(i);
            Log.d(TAG, String.format("Session %d = id:%d, startTime:%s, duration:%s, state:%s, returnCode:%s.",
                    i,
                    session.getSessionId(),
                    session.getStartTime(),
                    session.getDuration(),
                    session.getState(),
                    session.getReturnCode()));
        }
        Log.d(TAG, "Listed FFprobe sessions.");
    }

    public static void addUIAction(final Runnable runnable) {
        handler.post(runnable);
    }

    protected void registerApplicationFonts() throws IOException {
        final File cacheDirectory = getCacheDir();
        final File fontDirectory = new File(cacheDirectory, "fonts");

        boolean fontDirectoryCreated = fontDirectory.mkdirs();
        if (!fontDirectoryCreated) {
            android.util.Log.i(TAG, String.format("Failed to create font directory: %s.", fontDirectory.getAbsolutePath()));
        }

        // SAVE FONTS
        ResourcesUtil.rawResourceToFile(getResources(), R.raw.doppioone_regular, new File(fontDirectory, "doppioone_regular.ttf"));
        ResourcesUtil.rawResourceToFile(getResources(), R.raw.notosansarabic_regular, new File(fontDirectory, "notosansarabic_regular.ttf"));
        ResourcesUtil.rawResourceToFile(getResources(), R.raw.notosanssc_regular, new File(fontDirectory, "notosanssc_regular.ttf"));

        final HashMap<String, String> fontNameMapping = new HashMap<>();
        fontNameMapping.put("MyFontName", "Doppio One");
        FFmpegKitConfig.setFontDirectoryList(this, Arrays.asList(fontDirectory.getAbsolutePath(), "/system/fonts"), fontNameMapping);
        FFmpegKitConfig.setEnvironmentVariable("FFREPORT", String.format("file=%s", new File(cacheDirectory.getAbsolutePath(), "ffreport.txt").getAbsolutePath()));
    }

    protected void listSupportedCameraIds() {
        final List<String> supportedCameraIds = FFmpegKitConfig.getSupportedCameraIds(this);
        if (supportedCameraIds.isEmpty()) {
            android.util.Log.d(MainActivity.TAG, "No supported cameras found.");
        } else {
            for (String supportedCameraId : supportedCameraIds) {
                android.util.Log.d(MainActivity.TAG, "Supported camera detected: " + supportedCameraId);
            }
        }
    }

    static String notNull(final String string, final String valuePrefix) {
        return (string == null) ? "" : String.format("%s%s", valuePrefix, string);
    }

    private void setupTabs(final ViewPager2 viewPager, final PagerAdapter pagerAdapter) {
        final HorizontalScrollView tabScrollView = findViewById(R.id.tabScrollView);
        final LinearLayout tabStrip = findViewById(R.id.tabStrip);
        final float density = getResources().getDisplayMetrics().density;
        final int horizontalPadding = Math.round(24 * density);
        final int minTabWidth = Math.round(88 * density);
        final int selectedColor = ContextCompat.getColor(this, R.color.navigationColor);
        final int unselectedColor = Color.parseColor("#f39c12");

        for (int i = 0; i < pagerAdapter.getItemCount(); i++) {
            final int position = i;
            final TextView tab = new TextView(this);
            tab.setText(pagerAdapter.getPageTitle(position));
            tab.setGravity(android.view.Gravity.CENTER);
            tab.setSingleLine(true);
            tab.setTextColor(unselectedColor);
            tab.setTextSize(14);
            tab.setTypeface(Typeface.DEFAULT, Typeface.BOLD);
            tab.setPadding(horizontalPadding, 0, horizontalPadding, 0);
            tab.setMinWidth(minTabWidth);
            tab.setOnClickListener(v -> viewPager.setCurrentItem(position, true));
            tabStrip.addView(tab, new LinearLayout.LayoutParams(
                    LinearLayout.LayoutParams.WRAP_CONTENT,
                    LinearLayout.LayoutParams.MATCH_PARENT));
        }

        viewPager.registerOnPageChangeCallback(new ViewPager2.OnPageChangeCallback() {
            @Override
            public void onPageSelected(final int position) {
                updateSelectedTab(tabStrip, tabScrollView, position, selectedColor, unselectedColor);
            }
        });
        updateSelectedTab(tabStrip, tabScrollView, 0, selectedColor, unselectedColor);
    }

    private void updateSelectedTab(final LinearLayout tabStrip, final HorizontalScrollView tabScrollView, final int selectedPosition, final int selectedColor, final int unselectedColor) {
        for (int i = 0; i < tabStrip.getChildCount(); i++) {
            final TextView tab = (TextView) tabStrip.getChildAt(i);
            final boolean selected = i == selectedPosition;
            tab.setBackgroundResource(selected ? R.drawable.tab_indicator : 0);
            tab.setTextColor(selected ? selectedColor : unselectedColor);
        }

        final View selectedTab = tabStrip.getChildAt(selectedPosition);
        if (selectedTab != null) {
            selectedTab.post(() -> {
                final int scrollX = selectedTab.getLeft() - ((tabScrollView.getWidth() - selectedTab.getWidth()) / 2);
                tabScrollView.smoothScrollTo(scrollX, 0);
            });
        }
    }

}

File: android/test-app-kotlin/src/main/kotlin/com/arthenica/ffmpegkit/test/MainActivity.kt
/*
 * Copyright (c) 2018-2026 Taner Sener
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package com.arthenica.ffmpegkit.test

import android.content.Context
import android.graphics.Color
import android.graphics.Typeface
import android.media.MediaPlayer
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.util.Log
import android.widget.HorizontalScrollView
import android.widget.LinearLayout
import android.widget.MediaController
import android.widget.TextView
import android.widget.VideoView
import androidx.activity.SystemBarStyle
import androidx.activity.enableEdgeToEdge
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import androidx.core.graphics.toColorInt
import androidx.core.net.toUri
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import androidx.fragment.app.Fragment
import androidx.viewpager2.widget.ViewPager2
import com.arthenica.ffmpegkit.FFmpegKit
import com.arthenica.ffmpegkit.FFmpegKitConfig
import com.arthenica.ffmpegkit.FFprobeKit
import com.arthenica.ffmpegkit.Level
import com.arthenica.ffmpegkit.Signal
import com.arthenica.ffmpegkit.test.databinding.ActivityMainBinding
import com.arthenica.ffmpegkit.util.ResourcesUtil
import com.arthenica.smartexception.java.Exceptions
import java.io.File
import java.io.IOException
import kotlin.math.roundToInt

fun Fragment.cacheFile(name: String): File = File(requireContext().cacheDir, name)

fun Fragment.filesFile(name: String): File = File(requireContext().filesDir, name)

fun VideoView.playFile(context: Context, file: File) {
    val mediaController = MediaController(context)
    mediaController.setAnchorView(this)
    setVideoURI("file://${file.absolutePath}".toUri())
    setMediaController(mediaController)
    requestFocus()
    setOnPreparedListener { _: MediaPlayer -> setBackgroundColor(0x00000000) }
    setOnErrorListener { _: MediaPlayer, _: Int, _: Int ->
        stopPlayback()
        false
    }
    start()
}

class MainActivity : AppCompatActivity() {
    private lateinit var binding: ActivityMainBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        enableEdgeToEdge(SystemBarStyle.dark(Color.TRANSPARENT))

        super.onCreate(savedInstanceState)

        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)

        val pagerAdapter = PagerAdapter(this, this)
        binding.pager.adapter = pagerAdapter
        setupTabs(binding.pager, pagerAdapter)

        ViewCompat.setOnApplyWindowInsetsListener(binding.rootLayout) { _, windowInsets ->
            val bars = windowInsets.getInsets(
                WindowInsetsCompat.Type.systemBars() or WindowInsetsCompat.Type.displayCutout()
            )
            binding.toolbar.setPadding(0, bars.top, 0, 0)
            binding.pager.setPadding(bars.left, 0, bars.right, bars.bottom)
            WindowInsetsCompat.CONSUMED
        }

        try {
            registerApplicationFonts()
            Log.d(TAG, "Application fonts registered.")
        } catch (e: IOException) {
            Log.e(TAG, "Font registration failed.${Exceptions.getStackTraceString(e)}.")
        }

        Log.d(TAG, "Listing supported camera ids.")
        listSupportedCameraIds()

        FFmpegKitConfig.ignoreSignal(Signal.SIGXCPU)
        FFmpegKitConfig.setLogLevel(Level.AV_LOG_INFO)
    }

    @Throws(IOException::class)
    fun registerApplicationFonts() {
        val cacheDirectory = cacheDir
        val fontDirectory = File(cacheDirectory, "fonts")

        val fontDirectoryCreated = fontDirectory.mkdirs()
        if (!fontDirectoryCreated) {
            Log.i(TAG, "Failed to create font directory: ${fontDirectory.absolutePath}.")
        }

        ResourcesUtil.rawResourceToFile(resources, R.raw.doppioone_regular, File(fontDirectory, "doppioone_regular.ttf"))
        ResourcesUtil.rawResourceToFile(resources, R.raw.truenorg, File(fontDirectory, "truenorg.otf"))

        val fontNameMapping = hashMapOf<String?, String?>()
        fontNameMapping["MyFontName"] = "Doppio One"
        FFmpegKitConfig.setFontDirectoryList(this, listOf(fontDirectory.absolutePath, "/system/fonts"), fontNameMapping)
        FFmpegKitConfig.setEnvironmentVariable("FFREPORT", "file=${File(cacheDirectory.absolutePath, "ffreport.txt").absolutePath}")
    }

    fun listSupportedCameraIds() {
        val supportedCameraIds = FFmpegKitConfig.getSupportedCameraIds(this)
        if (supportedCameraIds.isEmpty()) {
            Log.d(TAG, "No supported cameras found.")
        } else {
            for (supportedCameraId in supportedCameraIds) {
                Log.d(TAG, "Supported camera detected: $supportedCameraId")
            }
        }
    }

    private fun setupTabs(viewPager: ViewPager2, pagerAdapter: PagerAdapter) {
        val tabScrollView = binding.tabScrollView
        val tabStrip = binding.tabStrip
        val density = resources.displayMetrics.density
        val horizontalPadding = (24 * density).roundToInt()
        val minTabWidth = (88 * density).roundToInt()
        val selectedColor = ContextCompat.getColor(this, R.color.navigationColor)
        val unselectedColor = "#f39c12".toColorInt()

        for (i in 0 until pagerAdapter.itemCount) {
            val tab = TextView(this)
            tab.text = pagerAdapter.getPageTitle(i)
            tab.gravity = android.view.Gravity.CENTER
            tab.setSingleLine(true)
            tab.setTextColor(unselectedColor)
            tab.textSize = 14f
            tab.setTypeface(Typeface.DEFAULT, Typeface.BOLD)
            tab.setPadding(horizontalPadding, 0, horizontalPadding, 0)
            tab.minWidth = minTabWidth
            tab.setOnClickListener { viewPager.setCurrentItem(i, true) }
            tabStrip.addView(
                tab,
                LinearLayout.LayoutParams(
                    LinearLayout.LayoutParams.WRAP_CONTENT,
                    LinearLayout.LayoutParams.MATCH_PARENT
                )
            )
        }

        viewPager.registerOnPageChangeCallback(object : ViewPager2.OnPageChangeCallback() {
            override fun onPageSelected(position: Int) {
                updateSelectedTab(tabStrip, tabScrollView, position, selectedColor, unselectedColor)
            }
        })
        updateSelectedTab(tabStrip, tabScrollView, 0, selectedColor, unselectedColor)
    }

    private fun updateSelectedTab(tabStrip: LinearLayout, tabScrollView: HorizontalScrollView, selectedPosition: Int, selectedColor: Int, unselectedColor: Int) {
        for (i in 0 until tabStrip.childCount) {
            val tab = tabStrip.getChildAt(i) as TextView
            val selected = i == selectedPosition
            tab.setBackgroundResource(if (selected) R.drawable.tab_indicator else 0)
            tab.setTextColor(if (selected) selectedColor else unselectedColor)
        }

        val selectedTab = tabStrip.getChildAt(selectedPosition)
        selectedTab?.post {
            val scrollX = selectedTab.left - (tabScrollView.width - selectedTab.width) / 2
            tabScrollView.smoothScrollTo(scrollX, 0)
        }
    }

    companion object {
        const val TAG = "ffmpeg-kit-next-test"

        init {
            Exceptions.registerRootPackage("com.arthenica")
        }

        @JvmField
        val handler = Handler(Looper.getMainLooper())

        @JvmStatic
        fun listFFmpegSessions() {
            val ffmpegSessions = FFmpegKit.listSessions()
            Log.d(TAG, "Listing FFmpeg sessions.")
            for (i in ffmpegSessions.indices) {
                val session = ffmpegSessions[i]
                Log.d(
                    TAG,
                    String.format(
                        "Session %d = id:%d, startTime:%s, duration:%s, state:%s, returnCode:%s.",
                        i,
                        session.getSessionId(),
                        session.getStartTime(),
                        session.getDuration(),
                        session.getState(),
                        session.getReturnCode()
                    )
                )
            }
            Log.d(TAG, "Listed FFmpeg sessions.")
        }

        @JvmStatic
        fun listFFprobeSessions() {
            val ffprobeSessions = FFprobeKit.listFFprobeSessions()
            Log.d(TAG, "Listing FFprobe sessions.")
            for (i in ffprobeSessions.indices) {
                val session = ffprobeSessions[i]
                Log.d(
                    TAG,
                    String.format(
                        "Session %d = id:%d, startTime:%s, duration:%s, state:%s, returnCode:%s.",
                        i,
                        session.getSessionId(),
                        session.getStartTime(),
                        session.getDuration(),
                        session.getState(),
                        session.getReturnCode()
                    )
                )
            }
            Log.d(TAG, "Listed FFprobe sessions.")
        }

        @JvmStatic
        fun addUIAction(action: () -> Unit) {
            handler.post(action)
        }

        @JvmStatic
        fun notNull(string: String?, valuePrefix: String): String {
            return if (string == null) "" else "$valuePrefix$string"
        }
    }
}
